File: documentation-examples-kotlin/src/kotlin/dev/typr/foundations/docs/analysis/QueryAnalysisNamed.kt
package dev.typr.foundations.docs.analysis

import dev.typr.foundationskt.*
import dev.typr.foundationskt.data.*
import java.sql.Connection

@Suppress("unused")
class QueryAnalysisNamed {
    data class User(val id: Int, val name: String, val email: String)

    private lateinit var connection: Connection
    private val userId = 1

    private val userRowParser: RowParser<User> =
        RowParser.builder<User>()
            .field(PgTypes.int4, User::id)
            .field(PgTypes.text, User::name)
            .field(PgTypes.text, User::email)
            .build(::User)

    //start
    fun analyzeNamedQuery() {
        val query =
            Sql { """
                SELECT id, name, email
                FROM users
                WHERE id = ${PgTypes.int4(userId)}
            """.trimIndent() }
                .query(userRowParser.all())

        // Give your query a name - it shows up in the error report
        val analysis =
            QueryAnalyzer.analyze("findUserById", query, connection)
                .single()

        if (!analysis.succeeded()) {
            throw AssertionError(analysis.report())
        }
    }
    //stop
}

File: documentation-examples-scala/src/scala/dev/typr/foundations/docs/analysis/QueryAnalysisNamed.scala
package dev.typr.foundations.docs.analysis
import dev.typr.foundationssc.*
import dev.typr.foundationssc.Fragment.*
import dev.typr.foundationssc.data.*

import java.sql.Connection

@SuppressWarnings(Array("unused"))
object QueryAnalysisNamed:
  case class User(id: Int, name: String, email: String)

  private val connection: Connection = null // placeholder
  private val userId = 1

  private val userRowParser: RowParser[User] = RowParser.builder[User]()
    .field(PgTypes.int4)(_.id)
    .field(PgTypes.text)(_.name)
    .field(PgTypes.text)(_.email)
    .build(User.apply)

  //start
  def analyzeNamedQuery(): Unit =
    val query =
      sql"""SELECT id, name, email
            FROM users
            WHERE id = ${PgTypes.int4(userId)}"""
        .query(userRowParser.all())

    // Give your query a name for the error report
    val analysis =
      QueryAnalyzer.analyze(
        "findUserById", query, connection
      ).head

    if !analysis.succeeded() then
      throw AssertionError(analysis.report())
  //stop
